bitcoin.go

```diff
@@ -70,43 +69,56 @@ func bitcoinDecodeToBytes(input string, config *CoinConfig) ([]byte, error) {
 		return nil, errors.New("invalid address")
 	}
 	// try base58 first
-	bytes, version, err := base58.CheckDecode(input)
+	data, err := Base58AddressDecodeToBytesPrefix(input, btcKeyHashLenght+len(config.P2SHPPrefix), [][]byte{config.P2PKHPrefix, config.P2SHPPrefix})
 	if err != nil {
 		if len(config.HRP) <= 0 {
 			return nil, err
 		}
 		// try bech32
-		decodedHrp, bytes, err := bech32.Decode(input)
-		if err != nil {
-			return nil, errors.Wrapf(err, "decoding bech32 failed")
+		decodedHrp, data, err2 := bech32.Decode(input)
+		if err2 != nil {
+			return nil, errors.Wrapf(err2, "decoding base58 and bech32 failed: %v", err.Error())
 		}
 		if decodedHrp != config.HRP {
 			return nil, errors.New("invalid hrp")
 		}
-		return buildWitnessScript(bytes)
+		return buildWitnessScript(data)
+	}
+	var prefix []byte
+	if len(data) >= len(config.P2PKHPrefix) {
+		prefix = data[:len(config.P2PKHPrefix)]
+		data = data[len(config.P2PKHPrefix):]
 	}
-	// check data length
-	if len(bytes) != btcKeyHashLenght {
-		return nil, errors.New("invalid data length")
+	// check prefix
+	if bytes.Equal(prefix, config.P2PKHPrefix) {
+		return buildP2PKHScript(data), nil
 	}
-	// check version byte
-	if version == config.P2PKHPrefix {
-		return buildP2PKHScript(bytes), nil
-	} else if version == config.P2SHPPrefix {
-		return buildP2SHScript(bytes), nil
+	// config.P2SHPPrefix)
+	return buildP2SHScript(data), nil
+}
+
+func replacePrefix(input []byte, oldPrefix []byte, newPrefix []byte) []byte {
```

Comment:
what does this method do?

Reply:
Removes a []byte prefix from input data if present, and adds another prefix instead.  Used to remove the opcode prefix like OP_DUP,OP_HASH160,20 and replace with P2P prefix like []byte{0x05}.  P2P prefix is one byte for most BTC forks, except ZCash, where it is 2.